// web3-eth-ens/dist/web3-eth-ens.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('web3-core-promievent'), require('web3-providers'), require('web3-utils'), require('web3-core-helpers'), require('web3-core-method'), require('web3-eth-abi'), require('web3-net'), require('web3-core'), require('@babel/runtime/regenerator'), require('@babel/runtime/helpers/asyncToGenerator'), require('@babel/runtime/helpers/classCallCheck'), require('@babel/runtime/helpers/createClass'), require('@babel/runtime/helpers/possibleConstructorReturn'), require('@babel/runtime/helpers/getPrototypeOf'), require('@babel/runtime/helpers/get'), require('@babel/runtime/helpers/inherits'), require('lodash/isFunction'), require('eth-ens-namehash'), require('web3-eth-contract')) :
    typeof define === 'function' && define.amd ? define(['exports', 'web3-core-promievent', 'web3-providers', 'web3-utils', 'web3-core-helpers', 'web3-core-method', 'web3-eth-abi', 'web3-net', 'web3-core', '@babel/runtime/regenerator', '@babel/runtime/helpers/asyncToGenerator', '@babel/runtime/helpers/classCallCheck', '@babel/runtime/helpers/createClass', '@babel/runtime/helpers/possibleConstructorReturn', '@babel/runtime/helpers/getPrototypeOf', '@babel/runtime/helpers/get', '@babel/runtime/helpers/inherits', 'lodash/isFunction', 'eth-ens-namehash', 'web3-eth-contract'], factory) :
    (factory((global.Web3EthEns = {}),global['web3-core-promievent'],global.web3Providers,global.Utils,global.web3CoreHelpers,global.web3CoreMethod,global.web3EthAbi,global.web3Net,global.web3Core,global._regeneratorRuntime,global._asyncToGenerator,global._classCallCheck,global._createClass,global._possibleConstructorReturn,global._getPrototypeOf,global._get,global._inherits,global.isFunction,global['eth-ens-namehash'],global['web3-eth-contract']));
}(this, (function (exports,web3CorePromievent,web3Providers,Utils,web3CoreHelpers,web3CoreMethod,web3EthAbi,web3Net,web3Core,_regeneratorRuntime,_asyncToGenerator,_classCallCheck,_createClass,_possibleConstructorReturn,_getPrototypeOf,_get,_inherits,isFunction,namehash,web3EthContract) { 'use strict';

    _regeneratorRuntime = _regeneratorRuntime && _regeneratorRuntime.hasOwnProperty('default') ? _regeneratorRuntime['default'] : _regeneratorRuntime;
    _asyncToGenerator = _asyncToGenerator && _asyncToGenerator.hasOwnProperty('default') ? _asyncToGenerator['default'] : _asyncToGenerator;
    _classCallCheck = _classCallCheck && _classCallCheck.hasOwnProperty('default') ? _classCallCheck['default'] : _classCallCheck;
    _createClass = _createClass && _createClass.hasOwnProperty('default') ? _createClass['default'] : _createClass;
    _possibleConstructorReturn = _possibleConstructorReturn && _possibleConstructorReturn.hasOwnProperty('default') ? _possibleConstructorReturn['default'] : _possibleConstructorReturn;
    _getPrototypeOf = _getPrototypeOf && _getPrototypeOf.hasOwnProperty('default') ? _getPrototypeOf['default'] : _getPrototypeOf;
    _get = _get && _get.hasOwnProperty('default') ? _get['default'] : _get;
    _inherits = _inherits && _inherits.hasOwnProperty('default') ? _inherits['default'] : _inherits;
    isFunction = isFunction && isFunction.hasOwnProperty('default') ? isFunction['default'] : isFunction;
    namehash = namehash && namehash.hasOwnProperty('default') ? namehash['default'] : namehash;

    var Ens =
    function (_AbstractWeb3Module) {
      _inherits(Ens, _AbstractWeb3Module);
      function Ens(provider, providersModuleFactory, methodModuleFactory, options, ensModuleFactory, contractModuleFactory, promiEvent, abiCoder, utils, formatters, registryOptions, net) {
        var _this;
        _classCallCheck(this, Ens);
        _this = _possibleConstructorReturn(this, _getPrototypeOf(Ens).call(this, provider, providersModuleFactory, methodModuleFactory, null, options));
        _this.ensModuleFactory = ensModuleFactory;
        _this.contractModuleFactory = contractModuleFactory;
        _this.promiEvent = promiEvent;
        _this.abiCoder = abiCoder;
        _this.utils = utils;
        _this.formatters = formatters;
        _this.registryOptions = registryOptions;
        _this.net = net;
        _this.transactionSigner = options.transactionSigner;
        _this._registry = false;
        return _this;
      }
      _createClass(Ens, [{
        key: "setProvider",
        value: function setProvider(provider, net) {
          return !!(_get(_getPrototypeOf(Ens.prototype), "setProvider", this).call(this, provider, net) && this.registry.setProvider(provider, net));
        }
      }, {
        key: "resolver",
        value: function resolver(name) {
          return this.registry.resolver(name);
        }
      }, {
        key: "supportsInterface",
        value: function () {
          var _supportsInterface = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee(name, interfaceId) {
            var callback,
                resolver,
                _args = arguments;
            return _regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    callback = _args.length > 2 && _args[2] !== undefined ? _args[2] : null;
                    _context.next = 3;
                    return this.registry.resolver(name);
                  case 3:
                    resolver = _context.sent;
                    return _context.abrupt("return", resolver.methods.supportsInterface(interfaceId).call(callback));
                  case 5:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));
          return function supportsInterface(_x, _x2) {
            return _supportsInterface.apply(this, arguments);
          };
        }()
      }, {
        key: "getAddress",
        value: function () {
          var _getAddress = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee2(name) {
            var callback,
                resolver,
                _args2 = arguments;
            return _regeneratorRuntime.wrap(function _callee2$(_context2) {
              while (1) {
                switch (_context2.prev = _context2.next) {
                  case 0:
                    callback = _args2.length > 1 && _args2[1] !== undefined ? _args2[1] : null;
                    _context2.next = 3;
                    return this.registry.resolver(name);
                  case 3:
                    resolver = _context2.sent;
                    return _context2.abrupt("return", resolver.methods.addr(namehash.hash(name)).call(callback));
                  case 5:
                  case "end":
                    return _context2.stop();
                }
              }
            }, _callee2, this);
          }));
          return function getAddress(_x3) {
            return _getAddress.apply(this, arguments);
          };
        }()
      }, {
        key: "setAddress",
        value: function setAddress(name, address, sendOptions) {
          var callback = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
          var promiEvent = new this.registry.PromiEvent();
          this.registry.resolver(name).then(function (resolver) {
            resolver.methods.setAddr(namehash.hash(name), address).send(sendOptions, callback).on('transactionHash', function (transactionHash) {
              promiEvent.emit('transactionHash', transactionHash);
            }).on('confirmation', function (confirmationNumber, receipt) {
              promiEvent.emit('confirmation', confirmationNumber, receipt);
            }).on('receipt', function (receipt) {
              if (isFunction(callback)) {
                callback(receipt);
              }
              promiEvent.emit('receipt', receipt);
              promiEvent.resolve(receipt);
            }).on('error', function (error) {
              if (isFunction(callback)) {
                callback(error);
              }
              promiEvent.emit('error', error);
              promiEvent.reject(error);
            });
          });
          return promiEvent;
        }
      }, {
        key: "getPubkey",
        value: function () {
          var _getPubkey = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee3(name) {
            var callback,
                resolver,
                _args3 = arguments;
            return _regeneratorRuntime.wrap(function _callee3$(_context3) {
              while (1) {
                switch (_context3.prev = _context3.next) {
                  case 0:
                    callback = _args3.length > 1 && _args3[1] !== undefined ? _args3[1] : null;
                    _context3.next = 3;
                    return this.registry.resolver(name);
                  case 3:
                    resolver = _context3.sent;
                    return _context3.abrupt("return", resolver.methods.pubkey(namehash.hash(name)).call(callback));
                  case 5:
                  case "end":
                    return _context3.stop();
                }
              }
            }, _callee3, this);
          }));
          return function getPubkey(_x4) {
            return _getPubkey.apply(this, arguments);
          };
        }()
      }, {
        key: "setPubkey",
        value: function setPubkey(name, x, y, sendOptions) {
          var callback = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : null;
          var promiEvent = new this.registry.PromiEvent();
          this.registry.resolver(name).then(function (resolver) {
            resolver.methods.setPubkey(namehash.hash(name), x, y).send(sendOptions, callback).on('transactionHash', function (transactionHash) {
              promiEvent.emit('transactionHash', transactionHash);
            }).on('confirmation', function (confirmationNumber, receipt) {
              promiEvent.emit('confirmation', confirmationNumber, receipt);
            }).on('receipt', function (receipt) {
              if (isFunction(callback)) {
                callback(receipt);
              }
              promiEvent.emit('receipt', receipt);
              promiEvent.resolve(receipt);
            }).on('error', function (error) {
              if (isFunction(callback)) {
                callback(error);
              }
              promiEvent.emit('error', error);
              promiEvent.reject(error);
            });
          });
          return promiEvent;
        }
      }, {
        key: "getText",
        value: function () {
          var _getText = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee4(name, key) {
            var callback,
                resolver,
                _args4 = arguments;
            return _regeneratorRuntime.wrap(function _callee4$(_context4) {
              while (1) {
                switch (_context4.prev = _context4.next) {
                  case 0:
                    callback = _args4.length > 2 && _args4[2] !== undefined ? _args4[2] : null;
                    _context4.next = 3;
                    return this.registry.resolver(name);
                  case 3:
                    resolver = _context4.sent;
                    return _context4.abrupt("return", resolver.methods.text(namehash.hash(name), key).call(callback));
                  case 5:
                  case "end":
                    return _context4.stop();
                }
              }
            }, _callee4, this);
          }));
          return function getText(_x5, _x6) {
            return _getText.apply(this, arguments);
          };
        }()
      }, {
        key: "setText",
        value: function setText(name, key, value, sendOptions) {
          var callback = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : null;
          var promiEvent = new this.registry.PromiEvent();
          this.registry.resolver(name).then(function (resolver) {
            resolver.methods.setText(namehash.hash(name), key, value).send(sendOptions, callback).on('transactionHash', function (transactionHash) {
              promiEvent.emit('transactionHash', transactionHash);
            }).on('confirmation', function (confirmationNumber, receipt) {
              promiEvent.emit('confirmation', confirmationNumber, receipt);
            }).on('receipt', function (receipt) {
              if (isFunction(callback)) {
                callback(receipt);
              }
              promiEvent.emit('receipt', receipt);
              promiEvent.resolve(receipt);
            }).on('error', function (error) {
              if (isFunction(callback)) {
                callback(error);
              }
              promiEvent.emit('error', error);
              promiEvent.reject(error);
            });
          });
          return promiEvent;
        }
      }, {
        key: "getContent",
        value: function () {
          var _getContent = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee5(name) {
            var callback,
                resolver,
                _args5 = arguments;
            return _regeneratorRuntime.wrap(function _callee5$(_context5) {
              while (1) {
                switch (_context5.prev = _context5.next) {
                  case 0:
                    callback = _args5.length > 1 && _args5[1] !== undefined ? _args5[1] : null;
                    _context5.next = 3;
                    return this.registry.resolver(name);
                  case 3:
                    resolver = _context5.sent;
                    return _context5.abrupt("return", resolver.methods.content(namehash.hash(name)).call(callback));
                  case 5:
                  case "end":
                    return _context5.stop();
                }
              }
            }, _callee5, this);
          }));
          return function getContent(_x7) {
            return _getContent.apply(this, arguments);
          };
        }()
      }, {
        key: "setContent",
        value: function setContent(name, hash, sendOptions) {
          var callback = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
          var promiEvent = new this.registry.PromiEvent();
          this.registry.resolver(name).then(function (resolver) {
            resolver.methods.setContent(namehash.hash(name), hash).send(sendOptions, callback).on('transactionHash', function (transactionHash) {
              promiEvent.emit('transactionHash', transactionHash);
            }).on('confirmation', function (confirmationNumber, receipt) {
              promiEvent.emit('confirmation', confirmationNumber, receipt);
            }).on('receipt', function (receipt) {
              if (isFunction(callback)) {
                callback(receipt);
              }
              promiEvent.emit('receipt', receipt);
              promiEvent.resolve(receipt);
            }).on('error', function (error) {
              if (isFunction(callback)) {
                callback(error);
              }
              promiEvent.emit('error', error);
              promiEvent.reject(error);
            });
          });
          return promiEvent;
        }
      }, {
        key: "getMultihash",
        value: function () {
          var _getMultihash = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee6(name) {
            var callback,
                resolver,
                _args6 = arguments;
            return _regeneratorRuntime.wrap(function _callee6$(_context6) {
              while (1) {
                switch (_context6.prev = _context6.next) {
                  case 0:
                    callback = _args6.length > 1 && _args6[1] !== undefined ? _args6[1] : null;
                    _context6.next = 3;
                    return this.registry.resolver(name);
                  case 3:
                    resolver = _context6.sent;
                    return _context6.abrupt("return", resolver.methods.multihash(namehash.hash(name)).call(callback));
                  case 5:
                  case "end":
                    return _context6.stop();
                }
              }
            }, _callee6, this);
          }));
          return function getMultihash(_x8) {
            return _getMultihash.apply(this, arguments);
          };
        }()
      }, {
        key: "setMultihash",
        value: function setMultihash(name, hash, sendOptions) {
          var callback = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
          var promiEvent = new this.registry.PromiEvent();
          this.registry.resolver(name).then(function (resolver) {
            resolver.methods.setMultihash(namehash.hash(name), hash).send(sendOptions, callback).on('transactionHash', function (transactionHash) {
              promiEvent.emit('transactionHash', transactionHash);
            }).on('confirmation', function (confirmationNumber, receipt) {
              promiEvent.emit('confirmation', confirmationNumber, receipt);
            }).on('receipt', function (receipt) {
              if (isFunction(callback)) {
                callback(receipt);
              }
              promiEvent.emit('receipt', receipt);
              promiEvent.resolve(receipt);
            }).on('error', function (error) {
              if (isFunction(callback)) {
                callback(error);
              }
              promiEvent.emit('error', error);
              promiEvent.reject(error);
            });
          });
          return promiEvent;
        }
      }, {
        key: "getContenthash",
        value: function () {
          var _getContenthash = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee7(name) {
            var callback,
                resolver,
                _args7 = arguments;
            return _regeneratorRuntime.wrap(function _callee7$(_context7) {
              while (1) {
                switch (_context7.prev = _context7.next) {
                  case 0:
                    callback = _args7.length > 1 && _args7[1] !== undefined ? _args7[1] : null;
                    _context7.next = 3;
                    return this.registry.resolver(name);
                  case 3:
                    resolver = _context7.sent;
                    return _context7.abrupt("return", resolver.methods.contenthash(namehash.hash(name)).call(callback));
                  case 5:
                  case "end":
                    return _context7.stop();
                }
              }
            }, _callee7, this);
          }));
          return function getContenthash(_x9) {
            return _getContenthash.apply(this, arguments);
          };
        }()
      }, {
        key: "setContenthash",
        value: function setContenthash(name, hash, sendOptions) {
          var callback = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
          var promiEvent = new this.registry.PromiEvent();
          this.registry.resolver(name).then(function (resolver) {
            resolver.methods.setContenthash(namehash.hash(name), hash).send(sendOptions, callback).on('transactionHash', function (transactionHash) {
              promiEvent.emit('transactionHash', transactionHash);
            }).on('confirmation', function (confirmationNumber, receipt) {
              promiEvent.emit('confirmation', confirmationNumber, receipt);
            }).on('receipt', function (receipt) {
              if (isFunction(callback)) {
                callback(receipt);
              }
              promiEvent.emit('receipt', receipt);
              promiEvent.resolve(receipt);
            }).on('error', function (error) {
              if (isFunction(callback)) {
                callback(error);
              }
              promiEvent.emit('error', error);
              promiEvent.reject(error);
            });
          });
          return promiEvent;
        }
      }, {
        key: "registry",
        get: function get() {
          if (!this._registry) {
            this._registry = this.ensModuleFactory.createRegistry(this.currentProvider, this.providersModuleFactory, this.methodModuleFactory, this.contractModuleFactory, this.promiEvent, this.abiCoder, this.utils, this.formatters, this.registryOptions, this.net);
          }
          return this._registry;
        }
      }]);
      return Ens;
    }(web3Core.AbstractWeb3Module);

    var REGISTRY_ABI = [{
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'resolver',
      outputs: [{
        name: '',
        type: 'address'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'owner',
      outputs: [{
        name: '',
        type: 'address'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'label',
        type: 'bytes32'
      }, {
        name: 'owner',
        type: 'address'
      }],
      name: 'setSubnodeOwner',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'ttl',
        type: 'uint64'
      }],
      name: 'setTTL',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'ttl',
      outputs: [{
        name: '',
        type: 'uint64'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'resolver',
        type: 'address'
      }],
      name: 'setResolver',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'owner',
        type: 'address'
      }],
      name: 'setOwner',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'owner',
        type: 'address'
      }],
      name: 'Transfer',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: true,
        name: 'label',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'owner',
        type: 'address'
      }],
      name: 'NewOwner',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'resolver',
        type: 'address'
      }],
      name: 'NewResolver',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'ttl',
        type: 'uint64'
      }],
      name: 'NewTTL',
      type: 'event'
    }];

    var RESOLVER_ABI = [{
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'key',
        type: 'string'
      }],
      name: 'text',
      outputs: [{
        name: '',
        type: 'string'
      }],
      payable: false,
      stateMutability: 'view',
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'key',
        type: 'string'
      }, {
        name: 'value',
        type: 'string'
      }],
      name: 'setText',
      outputs: [],
      payable: false,
      stateMutability: 'nonpayable',
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'interfaceID',
        type: 'bytes4'
      }],
      name: 'supportsInterface',
      outputs: [{
        name: '',
        type: 'bool'
      }],
      payable: false,
      stateMutability: 'pure',
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'contentTypes',
        type: 'uint256'
      }],
      name: 'ABI',
      outputs: [{
        name: 'contentType',
        type: 'uint256'
      }, {
        name: 'data',
        type: 'bytes'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'hash',
        type: 'bytes'
      }],
      name: 'setMultihash',
      outputs: [],
      payable: false,
      stateMutability: 'nonpayable',
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'multihash',
      outputs: [{
        name: '',
        type: 'bytes'
      }],
      payable: false,
      stateMutability: 'view',
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'hash',
        type: 'bytes'
      }],
      name: 'setContenthash',
      outputs: [],
      payable: false,
      stateMutability: 'nonpayable',
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'contenthash',
      outputs: [{
        name: '',
        type: 'bytes'
      }],
      payable: false,
      stateMutability: 'view',
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'x',
        type: 'bytes32'
      }, {
        name: 'y',
        type: 'bytes32'
      }],
      name: 'setPubkey',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'content',
      outputs: [{
        name: 'ret',
        type: 'bytes32'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'addr',
      outputs: [{
        name: 'ret',
        type: 'address'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'contentType',
        type: 'uint256'
      }, {
        name: 'data',
        type: 'bytes'
      }],
      name: 'setABI',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'name',
      outputs: [{
        name: 'ret',
        type: 'string'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'name',
        type: 'string'
      }],
      name: 'setName',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'hash',
        type: 'bytes32'
      }],
      name: 'setContent',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      constant: true,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }],
      name: 'pubkey',
      outputs: [{
        name: 'x',
        type: 'bytes32'
      }, {
        name: 'y',
        type: 'bytes32'
      }],
      payable: false,
      type: 'function'
    }, {
      constant: false,
      inputs: [{
        name: 'node',
        type: 'bytes32'
      }, {
        name: 'addr',
        type: 'address'
      }],
      name: 'setAddr',
      outputs: [],
      payable: false,
      type: 'function'
    }, {
      inputs: [{
        name: 'ensAddr',
        type: 'address'
      }],
      payable: false,
      type: 'constructor'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'a',
        type: 'address'
      }],
      name: 'AddrChanged',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'hash',
        type: 'bytes32'
      }],
      name: 'ContentChanged',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'indexedKey',
        type: 'string'
      }, {
        indexed: false,
        name: 'key',
        type: 'string'
      }],
      name: 'TextChanged',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'hash',
        type: 'bytes'
      }],
      name: 'ContenthashChanged',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'name',
        type: 'string'
      }],
      name: 'NameChanged',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: true,
        name: 'contentType',
        type: 'uint256'
      }],
      name: 'ABIChanged',
      type: 'event'
    }, {
      anonymous: false,
      inputs: [{
        indexed: true,
        name: 'node',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'x',
        type: 'bytes32'
      }, {
        indexed: false,
        name: 'y',
        type: 'bytes32'
      }],
      name: 'PubkeyChanged',
      type: 'event'
    }];

    var Registry =
    function (_AbstractContract) {
      _inherits(Registry, _AbstractContract);
      function Registry(provider, providersModuleFactory, methodModuleFactory, contractModuleFactory, promiEvent, abiCoder, utils, formatters, options, net) {
        var _this;
        _classCallCheck(this, Registry);
        _this = _possibleConstructorReturn(this, _getPrototypeOf(Registry).call(this, provider, providersModuleFactory, methodModuleFactory, contractModuleFactory, promiEvent, abiCoder, utils, formatters, REGISTRY_ABI, '', options));
        _this.net = net;
        _this.resolverContract = null;
        _this.resolverName = null;
        _this.checkNetwork().then(function (address) {
          _this.address = address;
        });
        return _this;
      }
      _createClass(Registry, [{
        key: "owner",
        value: function owner(name) {
          var _this2 = this;
          var callback = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
          return new Promise(function (resolve, reject) {
            _this2.methods.owner(namehash.hash(name)).call().then(function (receipt) {
              resolve(receipt);
              if (isFunction(callback)) {
                callback(false, receipt);
              }
            }).catch(function (error) {
              reject(error);
              if (isFunction(callback)) {
                callback(error, null);
              }
            });
          });
        }
      }, {
        key: "setProvider",
        value: function setProvider(provider, net) {
          if (this.resolverContract) {
            return !!(_get(_getPrototypeOf(Registry.prototype), "setProvider", this).call(this, provider, net) && this.resolverContract.setProvider(provider, net));
          }
          return _get(_getPrototypeOf(Registry.prototype), "setProvider", this).call(this, provider, net);
        }
      }, {
        key: "resolver",
        value: function () {
          var _resolver = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee(name) {
            var address, clone;
            return _regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    if (!(this.resolverName === name && this.resolverContract)) {
                      _context.next = 2;
                      break;
                    }
                    return _context.abrupt("return", this.resolverContract);
                  case 2:
                    _context.next = 4;
                    return this.methods.resolver(namehash.hash(name)).call();
                  case 4:
                    address = _context.sent;
                    clone = this.clone();
                    clone.jsonInterface = RESOLVER_ABI;
                    clone.address = address;
                    this.resolverName = name;
                    this.resolverContract = clone;
                    return _context.abrupt("return", clone);
                  case 11:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));
          return function resolver(_x) {
            return _resolver.apply(this, arguments);
          };
        }()
      }, {
        key: "checkNetwork",
        value: function () {
          var _checkNetwork = _asyncToGenerator(
          _regeneratorRuntime.mark(function _callee2() {
            var ensAddresses, block, headAge, networkType, address;
            return _regeneratorRuntime.wrap(function _callee2$(_context2) {
              while (1) {
                switch (_context2.prev = _context2.next) {
                  case 0:
                    ensAddresses = {
                      main: '0x314159265dD8dbb310642f98f50C066173C1259b',
                      ropsten: '0x112234455c3a32fd11230c42e7bccd4a84e02010',
                      rinkeby: '0xe7410170f87102df0055eb195163a03b7f2bff4a'
                    };
                    _context2.next = 3;
                    return this.net.getBlock('latest', false);
                  case 3:
                    block = _context2.sent;
                    headAge = new Date() / 1000 - block.timestamp;
                    if (!(headAge > 3600)) {
                      _context2.next = 7;
                      break;
                    }
                    throw new Error("Network not synced; last block was ".concat(headAge, " seconds ago"));
                  case 7:
                    _context2.next = 9;
                    return this.net.getNetworkType();
                  case 9:
                    networkType = _context2.sent;
                    address = ensAddresses[networkType];
                    if (!(typeof address === 'undefined')) {
                      _context2.next = 13;
                      break;
                    }
                    throw new TypeError("ENS is not supported on network: \"".concat(networkType, "\""));
                  case 13:
                    return _context2.abrupt("return", address);
                  case 14:
                  case "end":
                    return _context2.stop();
                }
              }
            }, _callee2, this);
          }));
          return function checkNetwork() {
            return _checkNetwork.apply(this, arguments);
          };
        }()
      }]);
      return Registry;
    }(web3EthContract.AbstractContract);

    var EnsModuleFactory =
    function () {
      function EnsModuleFactory() {
        _classCallCheck(this, EnsModuleFactory);
      }
      _createClass(EnsModuleFactory, [{
        key: "createENS",
        value: function createENS(provider, providersModuleFactory, methodModuleFactory, contractModuleFactory, promiEvent, abiCoder, utils, formatters, net, registryOptions, ensModuleOptions) {
          return new Ens(provider, providersModuleFactory, methodModuleFactory, ensModuleOptions, this, contractModuleFactory, promiEvent, abiCoder, utils, formatters, registryOptions, net);
        }
      }, {
        key: "createRegistry",
        value: function createRegistry(provider, providersModuleFactory, methodModuleFactory, contractModuleFactory, promiEvent, abiCoder, utils, formatters, options, net) {
          return new Registry(provider, providersModuleFactory, methodModuleFactory, contractModuleFactory, promiEvent, abiCoder, utils, formatters, options, net);
        }
      }]);
      return EnsModuleFactory;
    }();

    var Ens$1 = function Ens(provider, options, accounts) {
      var abiCoder = new web3EthAbi.AbiCoder();
      var methodModuleFactory = new web3CoreMethod.MethodModuleFactory();
      return new EnsModuleFactory().createENS(provider, new web3Providers.ProvidersModuleFactory(), new web3CoreMethod.MethodModuleFactory(), new web3EthContract.ContractModuleFactory(Utils, web3CoreHelpers.formatters, abiCoder, accounts, methodModuleFactory), web3CorePromievent.PromiEvent, abiCoder, Utils, web3CoreHelpers.formatters, new web3Net.Network(provider), {}, options);
    };

    exports.Ens = Ens$1;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
