// web3-eth/dist/web3-eth.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var web3CoreHelpers = require('web3-core-helpers');
var web3CoreSubscriptions = require('web3-core-subscriptions');
var web3EthAccounts = require('web3-eth-accounts');
var web3EthEns = require('web3-eth-ens');
var web3EthContract = require('web3-eth-contract');
var web3EthPersonal = require('web3-eth-personal');
var web3EthAbi = require('web3-eth-abi');
var web3EthIban = require('web3-eth-iban');
var web3Providers = require('web3-providers');
var web3Net = require('web3-net');
var Utils = require('web3-utils');
var _regeneratorRuntime = _interopDefault(require('@babel/runtime/regenerator'));
var _asyncToGenerator = _interopDefault(require('@babel/runtime/helpers/asyncToGenerator'));
var Nat = _interopDefault(require('eth-lib/lib/nat'));
var Bytes = _interopDefault(require('eth-lib/lib/bytes'));
var Hash = _interopDefault(require('eth-lib/lib/hash'));
var RLP = _interopDefault(require('eth-lib/lib/rlp'));
var Account = _interopDefault(require('eth-lib/lib/account'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _set = _interopDefault(require('@babel/runtime/helpers/set'));
var _get = _interopDefault(require('@babel/runtime/helpers/get'));
var web3Core = require('web3-core');
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var web3CoreMethod = require('web3-core-method');

var TransactionSigner =
function () {
  function TransactionSigner(utils, formatters) {
    _classCallCheck(this, TransactionSigner);
    this.utils = utils;
    this.formatters = formatters;
  }
  _createClass(TransactionSigner, [{
    key: "sign",
    value: function () {
      var _sign = _asyncToGenerator(
      _regeneratorRuntime.mark(function _callee(transaction, privateKey) {
        var rlpEncoded, hash, signature, rawTransaction, values;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (privateKey) {
                  _context.next = 2;
                  break;
                }
                throw new Error('No privateKey given to the TransactionSigner.');
              case 2:
                if (privateKey.startsWith('0x')) {
                  privateKey = privateKey.substring(2);
                }
                transaction = this.formatters.txInputFormatter(transaction);
                transaction.to = transaction.to || '0x';
                transaction.data = transaction.data || '0x';
                transaction.value = transaction.value || '0x';
                transaction.chainId = this.utils.numberToHex(transaction.chainId);
                rlpEncoded = this.createRlpEncodedTransaction(transaction);
                hash = Hash.keccak256(rlpEncoded);
                signature = this.createAccountSignature(hash, privateKey, transaction.chainId);
                rawTransaction = RLP.encode(this.mapRlpEncodedTransaction(rlpEncoded, signature));
                values = RLP.decode(rawTransaction);
                return _context.abrupt("return", {
                  messageHash: hash,
                  v: this.trimLeadingZero(values[6]),
                  r: this.trimLeadingZero(values[7]),
                  s: this.trimLeadingZero(values[8]),
                  rawTransaction: rawTransaction
                });
              case 14:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
      return function sign(_x, _x2) {
        return _sign.apply(this, arguments);
      };
    }()
  }, {
    key: "createRlpEncodedTransaction",
    value: function createRlpEncodedTransaction(transaction) {
      return RLP.encode([Bytes.fromNat(transaction.nonce), Bytes.fromNat(transaction.gasPrice), Bytes.fromNat(transaction.gas), transaction.to.toLowerCase(), Bytes.fromNat(transaction.value), transaction.data, Bytes.fromNat(transaction.chainId), '0x', '0x']);
    }
  }, {
    key: "createAccountSignature",
    value: function createAccountSignature(hash, privateKey, chainId) {
      return Account.makeSigner(Nat.toNumber(chainId) * 2 + 35)(hash, privateKey);
    }
  }, {
    key: "mapRlpEncodedTransaction",
    value: function mapRlpEncodedTransaction(rlpEncoded, signature) {
      var rawTransaction = RLP.decode(rlpEncoded).slice(0, 6).concat(Account.decodeSignature(signature));
      rawTransaction[6] = this.makeEven(this.trimLeadingZero(rawTransaction[6]));
      rawTransaction[7] = this.makeEven(this.trimLeadingZero(rawTransaction[7]));
      rawTransaction[8] = this.makeEven(this.trimLeadingZero(rawTransaction[8]));
      return rawTransaction;
    }
  }, {
    key: "trimLeadingZero",
    value: function trimLeadingZero(hex) {
      while (hex && hex.startsWith('0x0')) {
        hex = "0x".concat(hex.slice(3));
      }
      return hex;
    }
  }, {
    key: "makeEven",
    value: function makeEven(hex) {
      if (hex.length % 2 === 1) {
        hex = hex.replace('0x', '0x0');
      }
      return hex;
    }
  }]);
  return TransactionSigner;
}();

var Eth =
function (_AbstractWeb3Module) {
  _inherits(Eth, _AbstractWeb3Module);
  function Eth(provider, providersModuleFactory, methodModuleFactory, methodFactory, net, accounts, personal, Iban, abiCoder, ens, utils, formatters, subscriptionsFactory, contractModuleFactory, transactionSigner, options, nodeNet) {
    var _this;
    _classCallCheck(this, Eth);
    _this = _possibleConstructorReturn(this, _getPrototypeOf(Eth).call(this, provider, providersModuleFactory, methodModuleFactory, methodFactory, options, nodeNet));
    _this.net = net;
    _this.accounts = accounts;
    _this.personal = personal;
    _this.Iban = Iban;
    _this.abi = abiCoder;
    _this.ens = ens;
    _this.utils = utils;
    _this.formatters = formatters;
    _this.subscriptionsFactory = subscriptionsFactory;
    _this.contractModuleFactory = contractModuleFactory;
    _this.initiatedContracts = [];
    _this._transactionSigner = options.transactionSigner || transactionSigner;
    _this.Contract = function (abi, address) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      options.transactionSigner = _this.transactionSigner;
      var contract = _this.contractModuleFactory.createContract(_this.currentProvider, _this.providersModuleFactory, _this.accounts, abi, address, options);
      _this.initiatedContracts.push(contract);
      return contract;
    };
    return _this;
  }
  _createClass(Eth, [{
    key: "clearSubscriptions",
    value: function clearSubscriptions() {
      return _get(_getPrototypeOf(Eth.prototype), "clearSubscriptions", this).call(this, 'eth_unsubscribe');
    }
  }, {
    key: "subscribe",
    value: function subscribe(type, options, callback) {
      switch (type) {
        case 'logs':
          return this.subscriptionsFactory.createLogSubscription(options, this, this.methodFactory.createMethod('getPastLogs')).subscribe(callback);
        case 'newBlockHeaders':
          return this.subscriptionsFactory.createNewHeadsSubscription(this).subscribe(callback);
        case 'pendingTransactions':
          return this.subscriptionsFactory.createNewPendingTransactionsSubscription(this).subscribe(callback);
        case 'syncing':
          return this.subscriptionsFactory.createSyncingSubscription(this).subscribe(callback);
        default:
          throw new Error("Unknown subscription: ".concat(type));
      }
    }
  }, {
    key: "setProvider",
    value: function setProvider(provider, net) {
      var setContractProviders = this.initiatedContracts.every(function (contract) {
        return contract.setProvider(provider, net);
      });
      return this.net.setProvider(provider, net) && this.personal.setProvider(provider, net) && _get(_getPrototypeOf(Eth.prototype), "setProvider", this).call(this, provider, net) && setContractProviders;
    }
  }, {
    key: "transactionSigner",
    get: function get() {
      return this._transactionSigner;
    }
    ,
    set: function set(transactionSigner) {
      this._transactionSigner = transactionSigner;
      this.accounts.transactionSigner = transactionSigner;
      this.ens.transactionSigner = transactionSigner;
      this.initiatedContracts.forEach(function (contract) {
        contract.transactionSigner = transactionSigner;
      });
    }
  }, {
    key: "defaultGasPrice",
    set: function set(value) {
      this.initiatedContracts.forEach(function (contract) {
        contract.defaultGasPrice = value;
      });
      this.net.defaultGasPrice = value;
      this.personal.defaultGasPrice = value;
      _set(_getPrototypeOf(Eth.prototype), "defaultGasPrice", value, this, true);
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Eth.prototype), "defaultGasPrice", this);
    }
  }, {
    key: "defaultGas",
    set: function set(value) {
      this.initiatedContracts.forEach(function (contract) {
        contract.defaultGas = value;
      });
      this.net.defaultGas = value;
      this.personal.defaultGas = value;
      _set(_getPrototypeOf(Eth.prototype), "defaultGas", value, this, true);
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Eth.prototype), "defaultGas", this);
    }
  }, {
    key: "transactionBlockTimeout",
    set: function set(value) {
      this.initiatedContracts.forEach(function (contract) {
        contract.transactionBlockTimeout = value;
      });
      this.net.transactionBlockTimeout = value;
      this.personal.transactionBlockTimeout = value;
      _set(_getPrototypeOf(Eth.prototype), "transactionBlockTimeout", value, this, true);
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Eth.prototype), "transactionBlockTimeout", this);
    }
  }, {
    key: "transactionConfirmationBlocks",
    set: function set(value) {
      this.initiatedContracts.forEach(function (contract) {
        contract.transactionConfirmationBlocks = value;
      });
      this.net.transactionConfirmationBlocks = value;
      this.personal.transactionConfirmationBlocks = value;
      _set(_getPrototypeOf(Eth.prototype), "transactionConfirmationBlocks", value, this, true);
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Eth.prototype), "transactionConfirmationBlocks", this);
    }
  }, {
    key: "transactionPollingTimeout",
    set: function set(value) {
      this.initiatedContracts.forEach(function (contract) {
        contract.transactionPollingTimeout = value;
      });
      this.net.transactionPollingTimeout = value;
      this.personal.transactionPollingTimeout = value;
      _set(_getPrototypeOf(Eth.prototype), "transactionPollingTimeout", value, this, true);
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Eth.prototype), "transactionPollingTimeout", this);
    }
  }, {
    key: "defaultAccount",
    set: function set(value) {
      var _this2 = this;
      this.initiatedContracts.forEach(function (contract) {
        contract.defaultAccount = _this2.utils.toChecksumAddress(value);
      });
      this.net.defaultAccount = value;
      this.personal.defaultAccount = value;
      _set(_getPrototypeOf(Eth.prototype), "defaultAccount", value, this, true);
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Eth.prototype), "defaultAccount", this);
    }
  }, {
    key: "defaultBlock",
    set: function set(value) {
      this.initiatedContracts.forEach(function (contract) {
        contract.defaultBlock = value;
      });
      this.net.defaultBlock = value;
      this.personal.defaultBlock = value;
      _set(_getPrototypeOf(Eth.prototype), "defaultBlock", value, this, true);
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Eth.prototype), "defaultBlock", this);
    }
  }]);
  return Eth;
}(web3Core.AbstractWeb3Module);

var MethodFactory =
function (_AbstractMethodFactor) {
  _inherits(MethodFactory, _AbstractMethodFactor);
  function MethodFactory(methodModuleFactory, utils, formatters) {
    var _this;
    _classCallCheck(this, MethodFactory);
    _this = _possibleConstructorReturn(this, _getPrototypeOf(MethodFactory).call(this, methodModuleFactory, utils, formatters));
    _this.methods = {
      getNodeInfo: web3CoreMethod.GetNodeInfoMethod,
      getProtocolVersion: web3CoreMethod.GetProtocolVersionMethod,
      getCoinbase: web3CoreMethod.GetCoinbaseMethod,
      isMining: web3CoreMethod.IsMiningMethod,
      getHashrate: web3CoreMethod.GetHashrateMethod,
      isSyncing: web3CoreMethod.IsSyncingMethod,
      getGasPrice: web3CoreMethod.GetGasPriceMethod,
      getAccounts: web3CoreMethod.GetAccountsMethod,
      getBlockNumber: web3CoreMethod.GetBlockNumberMethod,
      getBalance: web3CoreMethod.GetBalanceMethod,
      getStorageAt: web3CoreMethod.GetStorageAtMethod,
      getCode: web3CoreMethod.GetCodeMethod,
      getBlock: web3CoreMethod.GetBlockMethod,
      getUncle: web3CoreMethod.GetUncleMethod,
      getBlockTransactionCount: web3CoreMethod.GetBlockTransactionCountMethod,
      getBlockUncleCount: web3CoreMethod.GetBlockUncleCountMethod,
      getTransaction: web3CoreMethod.GetTransactionMethod,
      getTransactionFromBlock: web3CoreMethod.GetTransactionFromBlockMethod,
      getTransactionReceipt: web3CoreMethod.GetTransactionReceipt,
      getTransactionCount: web3CoreMethod.GetTransactionCountMethod,
      sendSignedTransaction: web3CoreMethod.SendRawTransactionMethod,
      signTransaction: web3CoreMethod.SignTransactionMethod,
      sendTransaction: web3CoreMethod.SendTransactionMethod,
      sign: web3CoreMethod.SignMethod,
      call: web3CoreMethod.CallMethod,
      estimateGas: web3CoreMethod.EstimateGasMethod,
      submitWork: web3CoreMethod.SubmitWorkMethod,
      getWork: web3CoreMethod.GetWorkMethod,
      getPastLogs: web3CoreMethod.GetPastLogsMethod,
      requestAccounts: web3CoreMethod.RequestAccountsMethod,
      getId: web3CoreMethod.VersionMethod,
      getChainId: web3CoreMethod.ChainIdMethod
    };
    return _this;
  }
  return MethodFactory;
}(web3CoreMethod.AbstractMethodFactory);

var TransactionSigner$1 = function TransactionSigner$$1() {
  return new TransactionSigner(Utils, web3CoreHelpers.formatters);
};
var MethodFactory$1 = function MethodFactory$$1() {
  return new MethodFactory(new web3CoreMethod.MethodModuleFactory(), Utils, web3CoreHelpers.formatters);
};
var Eth$1 = function Eth$$1(provider, net, options) {
  var accounts = new web3EthAccounts.Accounts(provider, options);
  var abiCoder = new web3EthAbi.AbiCoder();
  var methodModuleFactory = new web3CoreMethod.MethodModuleFactory();
  return new Eth(provider, new web3Providers.ProvidersModuleFactory(), methodModuleFactory, new MethodFactory$1(), new web3Net.Network(provider, options), accounts, new web3EthPersonal.Personal(provider, accounts, options), web3EthIban.Iban, abiCoder, new web3EthEns.Ens(provider, accounts, options), Utils, web3CoreHelpers.formatters, new web3CoreSubscriptions.SubscriptionsFactory(), new web3EthContract.ContractModuleFactory(Utils, web3CoreHelpers.formatters, abiCoder, accounts, methodModuleFactory), new TransactionSigner$1(), options, net);
};

exports.TransactionSigner = TransactionSigner$1;
exports.MethodFactory = MethodFactory$1;
exports.Eth = Eth$1;
