// web3-shh/dist/web3-shh.esm.js
import { ProvidersModuleFactory } from 'web3-providers';
import { SubscriptionsFactory } from 'web3-core-subscriptions';
import { Network } from 'web3-net';
import * as Utils from 'web3-utils';
import { formatters } from 'web3-core-helpers';
import { AbstractWeb3Module } from 'web3-core';
import { AbstractMethodFactory, ShhVersionMethod, GetInfoMethod, SetMaxMessageSizeMethod, SetMinPoWMethod, MarkTrustedPeerMethod, NewKeyPairMethod, AddPrivateKeyMethod, DeleteKeyPairMethod, HasKeyPairMethod, GetPublicKeyMethod, GetPrivateKeyMethod, NewSymKeyMethod, AddSymKeyMethod, GenerateSymKeyFromPasswordMethod, HasSymKeyMethod, GetSymKeyMethod, DeleteSymKeyMethod, NewMessageFilterMethod, GetFilterMessagesMethod, DeleteMessageFilterMethod, PostMethod, MethodModuleFactory } from 'web3-core-method';

class Shh extends AbstractWeb3Module {
  constructor(provider, providersModuleFactory, methodModuleFactory, methodFactory, subscriptionsFactory, net, options, nodeNet) {
    super(provider, providersModuleFactory, methodModuleFactory, methodFactory, options, nodeNet);
    this.subscriptionsFactory = subscriptionsFactory;
    this.net = net;
  }
  subscribe(method, options, callback) {
    if (method === 'messages') {
      return this.subscriptionsFactory.createShhMessagesSubscription(options, this).subscribe(callback);
    }
    throw new Error(`Unknown subscription: ${method}`);
  }
  clearSubscriptions() {
    return super.clearSubscriptions('shh_unsubscribe');
  }
  setProvider(provider, net) {
    return super.setProvider(provider, net) && this.net.setProvider(provider, net);
  }
  set defaultGasPrice(value) {
    super.defaultGasPrice = value;
    this.net.defaultGasPrice = value;
  }
  get defaultGasPrice() {
    return super.defaultGasPrice;
  }
  set defaultGas(value) {
    super.defaultGas = value;
    this.net.defaultGas = value;
  }
  get defaultGas() {
    return super.defaultGas;
  }
  set transactionBlockTimeout(value) {
    super.transactionBlockTimeout = value;
    this.net.transactionBlockTimeout = value;
  }
  get transactionBlockTimeout() {
    return super.transactionBlockTimeout;
  }
  set transactionConfirmationBlocks(value) {
    super.transactionConfirmationBlocks = value;
    this.net.transactionConfirmationBlocks = value;
  }
  get transactionConfirmationBlocks() {
    return super.transactionConfirmationBlocks;
  }
  set transactionPollingTimeout(value) {
    super.transactionPollingTimeout = value;
    this.net.transactionPollingTimeout = value;
  }
  get transactionPollingTimeout() {
    return super.transactionPollingTimeout;
  }
  set defaultAccount(value) {
    super.defaultAccount = value;
    this.net.defaultAccount = value;
  }
  get defaultAccount() {
    return super.defaultAccount;
  }
  set defaultBlock(value) {
    super.defaultBlock = value;
    this.net.defaultBlock = value;
  }
  get defaultBlock() {
    return super.defaultBlock;
  }
}

class MethodFactory extends AbstractMethodFactory {
  constructor(methodModuleFactory, utils, formatters$$1) {
    super(methodModuleFactory, utils, formatters$$1);
    this.methods = {
      getVersion: ShhVersionMethod,
      getInfo: GetInfoMethod,
      setMaxMessageSize: SetMaxMessageSizeMethod,
      setMinPoW: SetMinPoWMethod,
      markTrustedPeer: MarkTrustedPeerMethod,
      newKeyPair: NewKeyPairMethod,
      addPrivateKey: AddPrivateKeyMethod,
      deleteKeyPair: DeleteKeyPairMethod,
      hasKeyPair: HasKeyPairMethod,
      getPublicKey: GetPublicKeyMethod,
      getPrivateKey: GetPrivateKeyMethod,
      newSymKey: NewSymKeyMethod,
      addSymKey: AddSymKeyMethod,
      generateSymKeyFromPassword: GenerateSymKeyFromPasswordMethod,
      hasSymKey: HasSymKeyMethod,
      getSymKey: GetSymKeyMethod,
      deleteSymKey: DeleteSymKeyMethod,
      newMessageFilter: NewMessageFilterMethod,
      getFilterMessages: GetFilterMessagesMethod,
      deleteMessageFilter: DeleteMessageFilterMethod,
      post: PostMethod
    };
  }
}

class ShhModuleFactory {
  constructor(utils, formatters$$1, methodModuleFactory) {
    this.utils = utils;
    this.formatters = formatters$$1;
    this.methodModuleFactory = methodModuleFactory;
  }
  createShhModule(provider, providersModuleFactory, subscriptionsFactory, net, options) {
    return new Shh(provider, providersModuleFactory, this.methodModuleFactory, this.createMethodFactory(), subscriptionsFactory, net, options);
  }
  createMethodFactory() {
    return new MethodFactory(this.methodModuleFactory, this.utils, this.formatters);
  }
}

const Shh$1 = (provider, net, options) => {
  return new ShhModuleFactory(Utils, formatters, new MethodModuleFactory()).createShhModule(provider, new ProvidersModuleFactory(), new SubscriptionsFactory(), new Network(provider, options), options, net);
};

export { Shh$1 as Shh };
