// web3-shh/dist/web3-shh.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var web3Providers = require('web3-providers');
var web3CoreSubscriptions = require('web3-core-subscriptions');
var web3Net = require('web3-net');
var Utils = require('web3-utils');
var web3CoreHelpers = require('web3-core-helpers');
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _set = _interopDefault(require('@babel/runtime/helpers/set'));
var _get = _interopDefault(require('@babel/runtime/helpers/get'));
var web3Core = require('web3-core');
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var web3CoreMethod = require('web3-core-method');

var Shh =
function (_AbstractWeb3Module) {
  _inherits(Shh, _AbstractWeb3Module);
  function Shh(provider, providersModuleFactory, methodModuleFactory, methodFactory, subscriptionsFactory, net, options, nodeNet) {
    var _this;
    _classCallCheck(this, Shh);
    _this = _possibleConstructorReturn(this, _getPrototypeOf(Shh).call(this, provider, providersModuleFactory, methodModuleFactory, methodFactory, options, nodeNet));
    _this.subscriptionsFactory = subscriptionsFactory;
    _this.net = net;
    return _this;
  }
  _createClass(Shh, [{
    key: "subscribe",
    value: function subscribe(method, options, callback) {
      if (method === 'messages') {
        return this.subscriptionsFactory.createShhMessagesSubscription(options, this).subscribe(callback);
      }
      throw new Error("Unknown subscription: ".concat(method));
    }
  }, {
    key: "clearSubscriptions",
    value: function clearSubscriptions() {
      return _get(_getPrototypeOf(Shh.prototype), "clearSubscriptions", this).call(this, 'shh_unsubscribe');
    }
  }, {
    key: "setProvider",
    value: function setProvider(provider, net) {
      return _get(_getPrototypeOf(Shh.prototype), "setProvider", this).call(this, provider, net) && this.net.setProvider(provider, net);
    }
  }, {
    key: "defaultGasPrice",
    set: function set(value) {
      _set(_getPrototypeOf(Shh.prototype), "defaultGasPrice", value, this, true);
      this.net.defaultGasPrice = value;
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Shh.prototype), "defaultGasPrice", this);
    }
  }, {
    key: "defaultGas",
    set: function set(value) {
      _set(_getPrototypeOf(Shh.prototype), "defaultGas", value, this, true);
      this.net.defaultGas = value;
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Shh.prototype), "defaultGas", this);
    }
  }, {
    key: "transactionBlockTimeout",
    set: function set(value) {
      _set(_getPrototypeOf(Shh.prototype), "transactionBlockTimeout", value, this, true);
      this.net.transactionBlockTimeout = value;
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Shh.prototype), "transactionBlockTimeout", this);
    }
  }, {
    key: "transactionConfirmationBlocks",
    set: function set(value) {
      _set(_getPrototypeOf(Shh.prototype), "transactionConfirmationBlocks", value, this, true);
      this.net.transactionConfirmationBlocks = value;
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Shh.prototype), "transactionConfirmationBlocks", this);
    }
  }, {
    key: "transactionPollingTimeout",
    set: function set(value) {
      _set(_getPrototypeOf(Shh.prototype), "transactionPollingTimeout", value, this, true);
      this.net.transactionPollingTimeout = value;
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Shh.prototype), "transactionPollingTimeout", this);
    }
  }, {
    key: "defaultAccount",
    set: function set(value) {
      _set(_getPrototypeOf(Shh.prototype), "defaultAccount", value, this, true);
      this.net.defaultAccount = value;
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Shh.prototype), "defaultAccount", this);
    }
  }, {
    key: "defaultBlock",
    set: function set(value) {
      _set(_getPrototypeOf(Shh.prototype), "defaultBlock", value, this, true);
      this.net.defaultBlock = value;
    }
    ,
    get: function get() {
      return _get(_getPrototypeOf(Shh.prototype), "defaultBlock", this);
    }
  }]);
  return Shh;
}(web3Core.AbstractWeb3Module);

var MethodFactory =
function (_AbstractMethodFactor) {
  _inherits(MethodFactory, _AbstractMethodFactor);
  function MethodFactory(methodModuleFactory, utils, formatters) {
    var _this;
    _classCallCheck(this, MethodFactory);
    _this = _possibleConstructorReturn(this, _getPrototypeOf(MethodFactory).call(this, methodModuleFactory, utils, formatters));
    _this.methods = {
      getVersion: web3CoreMethod.ShhVersionMethod,
      getInfo: web3CoreMethod.GetInfoMethod,
      setMaxMessageSize: web3CoreMethod.SetMaxMessageSizeMethod,
      setMinPoW: web3CoreMethod.SetMinPoWMethod,
      markTrustedPeer: web3CoreMethod.MarkTrustedPeerMethod,
      newKeyPair: web3CoreMethod.NewKeyPairMethod,
      addPrivateKey: web3CoreMethod.AddPrivateKeyMethod,
      deleteKeyPair: web3CoreMethod.DeleteKeyPairMethod,
      hasKeyPair: web3CoreMethod.HasKeyPairMethod,
      getPublicKey: web3CoreMethod.GetPublicKeyMethod,
      getPrivateKey: web3CoreMethod.GetPrivateKeyMethod,
      newSymKey: web3CoreMethod.NewSymKeyMethod,
      addSymKey: web3CoreMethod.AddSymKeyMethod,
      generateSymKeyFromPassword: web3CoreMethod.GenerateSymKeyFromPasswordMethod,
      hasSymKey: web3CoreMethod.HasSymKeyMethod,
      getSymKey: web3CoreMethod.GetSymKeyMethod,
      deleteSymKey: web3CoreMethod.DeleteSymKeyMethod,
      newMessageFilter: web3CoreMethod.NewMessageFilterMethod,
      getFilterMessages: web3CoreMethod.GetFilterMessagesMethod,
      deleteMessageFilter: web3CoreMethod.DeleteMessageFilterMethod,
      post: web3CoreMethod.PostMethod
    };
    return _this;
  }
  return MethodFactory;
}(web3CoreMethod.AbstractMethodFactory);

var ShhModuleFactory =
function () {
  function ShhModuleFactory(utils, formatters, methodModuleFactory) {
    _classCallCheck(this, ShhModuleFactory);
    this.utils = utils;
    this.formatters = formatters;
    this.methodModuleFactory = methodModuleFactory;
  }
  _createClass(ShhModuleFactory, [{
    key: "createShhModule",
    value: function createShhModule(provider, providersModuleFactory, subscriptionsFactory, net, options) {
      return new Shh(provider, providersModuleFactory, this.methodModuleFactory, this.createMethodFactory(), subscriptionsFactory, net, options);
    }
  }, {
    key: "createMethodFactory",
    value: function createMethodFactory() {
      return new MethodFactory(this.methodModuleFactory, this.utils, this.formatters);
    }
  }]);
  return ShhModuleFactory;
}();

var Shh$1 = function Shh(provider, net, options) {
  return new ShhModuleFactory(Utils, web3CoreHelpers.formatters, new web3CoreMethod.MethodModuleFactory()).createShhModule(provider, new web3Providers.ProvidersModuleFactory(), new web3CoreSubscriptions.SubscriptionsFactory(), new web3Net.Network(provider, options), options, net);
};

exports.Shh = Shh$1;
