// web3-shh/dist/web3-shh.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('web3-providers'), require('web3-core-subscriptions'), require('web3-net'), require('web3-utils'), require('web3-core-helpers'), require('@babel/runtime/helpers/createClass'), require('@babel/runtime/helpers/set'), require('@babel/runtime/helpers/get'), require('web3-core'), require('@babel/runtime/helpers/classCallCheck'), require('@babel/runtime/helpers/possibleConstructorReturn'), require('@babel/runtime/helpers/getPrototypeOf'), require('@babel/runtime/helpers/inherits'), require('web3-core-method')) :
    typeof define === 'function' && define.amd ? define(['exports', 'web3-providers', 'web3-core-subscriptions', 'web3-net', 'web3-utils', 'web3-core-helpers', '@babel/runtime/helpers/createClass', '@babel/runtime/helpers/set', '@babel/runtime/helpers/get', 'web3-core', '@babel/runtime/helpers/classCallCheck', '@babel/runtime/helpers/possibleConstructorReturn', '@babel/runtime/helpers/getPrototypeOf', '@babel/runtime/helpers/inherits', 'web3-core-method'], factory) :
    (factory((global.Web3Shh = {}),global.web3Providers,global.web3CoreSubscriptions,global.web3Net,global.Utils,global.web3CoreHelpers,global._createClass,global._set,global._get,global.web3Core,global._classCallCheck,global._possibleConstructorReturn,global._getPrototypeOf,global._inherits,global.web3CoreMethod));
}(this, (function (exports,web3Providers,web3CoreSubscriptions,web3Net,Utils,web3CoreHelpers,_createClass,_set,_get,web3Core,_classCallCheck,_possibleConstructorReturn,_getPrototypeOf,_inherits,web3CoreMethod) { 'use strict';

    _createClass = _createClass && _createClass.hasOwnProperty('default') ? _createClass['default'] : _createClass;
    _set = _set && _set.hasOwnProperty('default') ? _set['default'] : _set;
    _get = _get && _get.hasOwnProperty('default') ? _get['default'] : _get;
    _classCallCheck = _classCallCheck && _classCallCheck.hasOwnProperty('default') ? _classCallCheck['default'] : _classCallCheck;
    _possibleConstructorReturn = _possibleConstructorReturn && _possibleConstructorReturn.hasOwnProperty('default') ? _possibleConstructorReturn['default'] : _possibleConstructorReturn;
    _getPrototypeOf = _getPrototypeOf && _getPrototypeOf.hasOwnProperty('default') ? _getPrototypeOf['default'] : _getPrototypeOf;
    _inherits = _inherits && _inherits.hasOwnProperty('default') ? _inherits['default'] : _inherits;

    var Shh =
    function (_AbstractWeb3Module) {
      _inherits(Shh, _AbstractWeb3Module);
      function Shh(provider, providersModuleFactory, methodModuleFactory, methodFactory, subscriptionsFactory, net, options, nodeNet) {
        var _this;
        _classCallCheck(this, Shh);
        _this = _possibleConstructorReturn(this, _getPrototypeOf(Shh).call(this, provider, providersModuleFactory, methodModuleFactory, methodFactory, options, nodeNet));
        _this.subscriptionsFactory = subscriptionsFactory;
        _this.net = net;
        return _this;
      }
      _createClass(Shh, [{
        key: "subscribe",
        value: function subscribe(method, options, callback) {
          if (method === 'messages') {
            return this.subscriptionsFactory.createShhMessagesSubscription(options, this).subscribe(callback);
          }
          throw new Error("Unknown subscription: ".concat(method));
        }
      }, {
        key: "clearSubscriptions",
        value: function clearSubscriptions() {
          return _get(_getPrototypeOf(Shh.prototype), "clearSubscriptions", this).call(this, 'shh_unsubscribe');
        }
      }, {
        key: "setProvider",
        value: function setProvider(provider, net) {
          return _get(_getPrototypeOf(Shh.prototype), "setProvider", this).call(this, provider, net) && this.net.setProvider(provider, net);
        }
      }, {
        key: "defaultGasPrice",
        set: function set(value) {
          _set(_getPrototypeOf(Shh.prototype), "defaultGasPrice", value, this, true);
          this.net.defaultGasPrice = value;
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Shh.prototype), "defaultGasPrice", this);
        }
      }, {
        key: "defaultGas",
        set: function set(value) {
          _set(_getPrototypeOf(Shh.prototype), "defaultGas", value, this, true);
          this.net.defaultGas = value;
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Shh.prototype), "defaultGas", this);
        }
      }, {
        key: "transactionBlockTimeout",
        set: function set(value) {
          _set(_getPrototypeOf(Shh.prototype), "transactionBlockTimeout", value, this, true);
          this.net.transactionBlockTimeout = value;
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Shh.prototype), "transactionBlockTimeout", this);
        }
      }, {
        key: "transactionConfirmationBlocks",
        set: function set(value) {
          _set(_getPrototypeOf(Shh.prototype), "transactionConfirmationBlocks", value, this, true);
          this.net.transactionConfirmationBlocks = value;
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Shh.prototype), "transactionConfirmationBlocks", this);
        }
      }, {
        key: "transactionPollingTimeout",
        set: function set(value) {
          _set(_getPrototypeOf(Shh.prototype), "transactionPollingTimeout", value, this, true);
          this.net.transactionPollingTimeout = value;
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Shh.prototype), "transactionPollingTimeout", this);
        }
      }, {
        key: "defaultAccount",
        set: function set(value) {
          _set(_getPrototypeOf(Shh.prototype), "defaultAccount", value, this, true);
          this.net.defaultAccount = value;
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Shh.prototype), "defaultAccount", this);
        }
      }, {
        key: "defaultBlock",
        set: function set(value) {
          _set(_getPrototypeOf(Shh.prototype), "defaultBlock", value, this, true);
          this.net.defaultBlock = value;
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Shh.prototype), "defaultBlock", this);
        }
      }]);
      return Shh;
    }(web3Core.AbstractWeb3Module);

    var MethodFactory =
    function (_AbstractMethodFactor) {
      _inherits(MethodFactory, _AbstractMethodFactor);
      function MethodFactory(methodModuleFactory, utils, formatters) {
        var _this;
        _classCallCheck(this, MethodFactory);
        _this = _possibleConstructorReturn(this, _getPrototypeOf(MethodFactory).call(this, methodModuleFactory, utils, formatters));
        _this.methods = {
          getVersion: web3CoreMethod.ShhVersionMethod,
          getInfo: web3CoreMethod.GetInfoMethod,
          setMaxMessageSize: web3CoreMethod.SetMaxMessageSizeMethod,
          setMinPoW: web3CoreMethod.SetMinPoWMethod,
          markTrustedPeer: web3CoreMethod.MarkTrustedPeerMethod,
          newKeyPair: web3CoreMethod.NewKeyPairMethod,
          addPrivateKey: web3CoreMethod.AddPrivateKeyMethod,
          deleteKeyPair: web3CoreMethod.DeleteKeyPairMethod,
          hasKeyPair: web3CoreMethod.HasKeyPairMethod,
          getPublicKey: web3CoreMethod.GetPublicKeyMethod,
          getPrivateKey: web3CoreMethod.GetPrivateKeyMethod,
          newSymKey: web3CoreMethod.NewSymKeyMethod,
          addSymKey: web3CoreMethod.AddSymKeyMethod,
          generateSymKeyFromPassword: web3CoreMethod.GenerateSymKeyFromPasswordMethod,
          hasSymKey: web3CoreMethod.HasSymKeyMethod,
          getSymKey: web3CoreMethod.GetSymKeyMethod,
          deleteSymKey: web3CoreMethod.DeleteSymKeyMethod,
          newMessageFilter: web3CoreMethod.NewMessageFilterMethod,
          getFilterMessages: web3CoreMethod.GetFilterMessagesMethod,
          deleteMessageFilter: web3CoreMethod.DeleteMessageFilterMethod,
          post: web3CoreMethod.PostMethod
        };
        return _this;
      }
      return MethodFactory;
    }(web3CoreMethod.AbstractMethodFactory);

    var ShhModuleFactory =
    function () {
      function ShhModuleFactory(utils, formatters, methodModuleFactory) {
        _classCallCheck(this, ShhModuleFactory);
        this.utils = utils;
        this.formatters = formatters;
        this.methodModuleFactory = methodModuleFactory;
      }
      _createClass(ShhModuleFactory, [{
        key: "createShhModule",
        value: function createShhModule(provider, providersModuleFactory, subscriptionsFactory, net, options) {
          return new Shh(provider, providersModuleFactory, this.methodModuleFactory, this.createMethodFactory(), subscriptionsFactory, net, options);
        }
      }, {
        key: "createMethodFactory",
        value: function createMethodFactory() {
          return new MethodFactory(this.methodModuleFactory, this.utils, this.formatters);
        }
      }]);
      return ShhModuleFactory;
    }();

    var Shh$1 = function Shh(provider, net, options) {
      return new ShhModuleFactory(Utils, web3CoreHelpers.formatters, new web3CoreMethod.MethodModuleFactory()).createShhModule(provider, new web3Providers.ProvidersModuleFactory(), new web3CoreSubscriptions.SubscriptionsFactory(), new web3Net.Network(provider, options), options, net);
    };

    exports.Shh = Shh$1;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
